packages/apex-node/src/utils/traceFlags.ts

```diff
@@ -0,0 +1,233 @@
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private async findDebugLevel(
+    debugLevelName: string
+  ): Promise<string | undefined> {
+    const query = util.format(
+      "SELECT Id FROM DebugLevel WHERE DeveloperName = '%s'",
+      debugLevelName
+    );
+    const result = (await this.connection.tooling.query(query)) as QueryRecord;
+    return result.totalSize && result.totalSize > 0 && result.records
+      ? result.records[0].Id
+      : undefined;
+  }
+
+  private async updateDebugLevel(id: string): Promise<boolean> {
+    const debugLevel = {
+      Id: id,
+      ApexCode: 'FINEST',
+      Visualforce: 'FINER'
+    };
+    const result = (await this.connection.tooling.update(
+      'DebugLevel',
+      debugLevel
+    )) as DataRecordResult;
+    return result.success;
+  }
+
+  private async createDebugLevel(
+    debugLevelName: string
+  ): Promise<string | undefined> {
+    const developerName = debugLevelName;
+    const debugLevel = {
+      developerName,
+      MasterLabel: developerName,
+      ApexCode: 'FINEST',
+      Visualforce: 'FINER'
+    };
+    const result = (await this.connection.tooling.create(
+      'DebugLevel',
+      debugLevel
+    )) as DataRecordResult;
+    return result.success && result.id ? result.id : undefined;
+  }
+
+  private async updateTraceFlag(
+    id: string,
+    expirationDate: Date
+  ): Promise<boolean> {
+    const traceFlag = {
+      Id: id,
+      StartDate: '',
+      ExpirationDate: expirationDate.toUTCString()
+    };
+    const result = (await this.connection.tooling.update(
+      'TraceFlag',
+      traceFlag
+    )) as DataRecordResult;
+    return result.success;
+  }
+
+  private async createTraceFlag(
+    userId: string,
+    debugLevelId: string,
+    expirationDate: Date
+  ): Promise<string | undefined> {
+    const traceFlag = {
+      tracedentityid: userId,
+      logtype: LOG_TYPE,
+      debuglevelid: debugLevelId,
+      StartDate: '',
+      ExpirationDate: expirationDate.toUTCString()
+    };
+
+    const result = (await this.connection.tooling.create(
+      'TraceFlag',
+      traceFlag
+    )) as DataRecordResult;
+    return result.success && result.id ? result.id : undefined;
+  }
+
+  private isValidDateLength(expirationDate: Date): boolean {
+    const currDate = new Date().valueOf();
+    return (
+      expirationDate.getTime() - currDate >
+      this.LOG_TIMER_LENGTH_MINUTES * this.MILLISECONDS_PER_MINUTE
+    );
+  }
+
+  private calculateExpirationDate(expirationDate: Date): Date {
+    if (!this.isValidDateLength(expirationDate)) {
+      expirationDate = new Date(
+        Date.now() +
+          this.LOG_TIMER_LENGTH_MINUTES * this.MILLISECONDS_PER_MINUTE
+      );
+    }
+    return expirationDate;
+  }
+
+  private async getUserIdOrThrow(username: string): Promise<UserRecord> {
+    const userQuery = `SELECT id FROM User WHERE username='${username}'`;
```

Comment:
can we check or escape the query here to prevent injection attacks? https://developer.salesforce.com/docs/atlas.en-us.pages.meta/pages/pages_security_tips_soql_injection.htm

Reply:
Ah, good eye @xyc! I certainly can, but I don't think we need to here. We get the username from the connection, which comes directly from Salesforce. Is there a concern this could possibly be tampered with? If so, I'll go ahead and escape the username variable 👍

Reply:
The salesforce username can contain a single quotation mark - for example: `xchen+test'@salesforce.com` is a valid username.

Reply:
Good point - I'll update this!

Reply:
Done! Escaped the username and the debug level. Let me know if you're ok with where I added the escape; I didn't like having the dependency of reaching up to the execute directory, but I also didn't feel like making a util/utils.ts. I can do that though, if you think that would be better here 👍